### PR TITLE
Spruce - Angela F.

### PR DESCRIPTION
**Pull request by fan-gela**

### CODE REVIEW

File: viewing_party/party.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
+# Helper Functions
+def user_has_watched(movie, watched_movies):
+    if movie in watched_movies:
+        return True
+    else:
+        return False
```
Comment (anselrognlie) [inline viewing_party/party.py:6]: Great to see some custom helper functions!

In python, we don't typically do a boolean check, and then return a boolean. Usually, we return (or use directly) the result of the initial check (occasionally may need to invert it with `not`). In this case:
```python
def user_has_watched(movie, watched_movies):
    return movie in watched_movies
```

As a result, you might decide not to have this helper (since calling the helper is longer than the equivalent expression). But there are other cases where you might decide to have a helper because it provides a concrete name to an abstract operation. For example, incrementing a value is an abstract operation, but increasing a speed would be more concrete, and could be used to make code more self-descriptive.

But in this case, `movie in watched_movies` seems pretty clear to me.

Notice that for the next function, the implementation would be the same, but most likely, you would have good variable names in the scope where you were calling these functions. So one call would probably simply down to `movie in user_watched_movies` and the other, `movie in friends_watched_movies`. Both read pretty well to me.

File: viewing_party/party.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    else:
+        return False 
+    
+def in_friends_watchlist(movie, watched_by_friends):
+    if movie in watched_by_friends:
+        return True
+    else:
+        return False
+
+def get_friends_movies(user_data):
```
Comment (anselrognlie) [inline viewing_party/party.py:14]: Really nice way to use an auxiliary set to track of bit of immutable data from a mutable object to help you build up a unique list of objects efficiently!

It might be nice to include something about the uniqueness of the result in the function name to help reasoning about code in the places you call this function!

File: viewing_party/party.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+    if movie_title and genre and rating:
+        movie_dict = {
+            'title' : movie_title, 
+            'genre' : genre, 
+            'rating' : rating
+        }
+        return movie_dict
+    else: 
+        return None
```
Comment (anselrognlie) [inline viewing_party/party.py:36]: Consider reversing the sense of your if condition here, converting it into what is often called a guard clause, or guard condition. Guard clauses are used to check some trivial conditions that mean we wouldn't have to do any additional work in the function (or might even cause an error if we tried) and bail out quickly. In this case, it could look like
```python
def create_movie(movie_title, genre, rating):
    if not movie_title or not genre or not rating:
        return None
        
    movie_dict = {
        'title': movie_title, 
        'genre': genre, 
        'rating': rating
    }
    return movie_dict
```
Notice how this lets us reduce a layer if indentation around the _important_ code, and avoids that little dangling else at the end (which can be easy to overlook).

File: viewing_party/party.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+def add_to_watched(user_data, movie):
```
Comment (anselrognlie) [inline viewing_party/party.py:38]: 👍

File: viewing_party/party.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    watched_list = user_data["watched"]
+    watched_list.append(movie)
+    return user_data
+
+def add_to_watchlist(user_data, movie):
```
Comment (anselrognlie) [inline viewing_party/party.py:43]: 👍

File: viewing_party/party.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    add_watchlist = user_data["watchlist"]
+    add_watchlist.append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    movies_watched = user_data["watched"]
+    movies_to_watch = user_data["watchlist"]
+
+    for movie in list(movies_to_watch):
```
Comment (anselrognlie) [inline viewing_party/party.py:52]: Really clever to make an in-line copy of the list of movies to watch just before iterating over it. This means that when we remove the found movie in the body of the loop, the list we are modifying isn't the same as the list over which we are iterating. Modifying a list during iteration can lead to unexpected results. In our case it, it turns out not to be a problem practically, but it's great to see you thinking about this!

File: viewing_party/party.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        if title == movie["title"]:
+            movies_watched.append(movie)
+            movies_to_watch.remove(movie)
+    return user_data
+
+
+# Wave 2
+def get_watched_avg_rating(user_data):
+    
+    avg_rating_list = [a_dict['rating'] for a_dict in user_data['watched']]
```
Comment (anselrognlie) [inline viewing_party/party.py:62]: Nice comprehension to pick out the ratings of the user watched movies. Even in a small expression like this, it's good to use descriptive names for the things we're working with. While the movies _are_ dictionaries (and so a_dict describes the type accurately), the dictionary itself represents a movie, arguable just as important (or even more-so). I would tend to prefer `movie` as the variable in the comprehension. Or if you'd lie to include the type information, `movie_dict`.

File: viewing_party/party.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+    list_length = len(avg_rating_list)
+    list_total = sum(avg_rating_list)
+    
+    if list_length == 0:
```
Comment (anselrognlie) [inline viewing_party/party.py:67]: A more pythonic way to check whether a list is empty is
```python
    if not avg_rating_list:
```

File: viewing_party/party.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        return float(0)
+    else:
+        avg_rating = list_total/list_length
+        return avg_rating
+
+def get_most_watched_genre(user_data):
+
+    popular_genre_list = [a_dict['genre'] for a_dict in user_data['watched']]
+    most_common_genre = max(popular_genre_list, key = popular_genre_list.count, default = None)
```
Comment (anselrognlie) [inline viewing_party/party.py:76]: In the review, we talked about how `max(popular_genre_dict, key= popular_genre_dict.get)` would work. Based on that, can you describe how this works? Can you describe the inner workings of `max` and how it uses the `key` argument? Could you write this function if needed?

It's great to see folks trying new approaches, but if you try something _really_ outside the curriculum, it's nice to include a comment about what research you did and how it works. We don't want to incorporate code snippets into our code that we don't understand. How would we debug something that we don't understand why it does what it does? So aim for understanding in everything you write!

File: viewing_party/party.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    
+    return most_common_genre
+
+# Wave 3
+def get_unique_watched(user_data):
+    
+    friends_movies = get_friends_movies(user_data)
+    friend_movie_recs = []
+    for movie in user_data['watched']:
+        if movie not in friends_movies:
```
Comment (anselrognlie) [inline viewing_party/party.py:86]: Using `in` to check for list membership is a O(n) operation. In the same way that you used a supplementary list to build a set of titles in your helper method, you could do something similar here to look up the titles of the user movies in O(1) time. When working with lists, we need to be careful of hidden loops in some of the helpers it provides!

File: viewing_party/party.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+            friend_movie_recs.append(movie)
+
+    if len(user_data['watched']) == None:
```
Comment (anselrognlie) [inline viewing_party/party.py:89]: Notice that this condition will never be true, and the `return 0` will never run. `len()` returns a number, and a number is never equal to None. We don't really need this check anyway, since the expected test behavior when there are no unique user movies is an empty list, not None.

File: viewing_party/party.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+def get_friends_unique_watched(user_data):
+    friends_movies = get_friends_movies(user_data)
+
+    user_need_to_watch_list = []
+    for movie in friends_movies:
+        if movie not in user_data['watched']:
+            user_need_to_watch_list.append(movie)
+
+    user_need_to_watch_list_no_dup = []
+    [user_need_to_watch_list_no_dup.append(x) for x in user_need_to_watch_list if x not in user_need_to_watch_list_no_dup]
```
Comment (anselrognlie) [inline viewing_party/party.py:103]: We shouldn't use list comprehensions as a shortcut for iteration. We should only use comprehensions if we are actually building a new list. While this code _does_ iterate over `user_need_to_watch_list` and appends unique things to the no dups list, it will also end up creating a temporary list, the same length as the number of items in `user_need_to_watch_list`, filled with None (since append doesn't return a value) and then throws it a way.

File: viewing_party/party.py [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+def get_available_recs(user_data):
+    friends_movies = get_friends_movies(user_data)
+    movie_recs = []
+
+    if len(friends_movies) == 0:
+        return []
+
+    titles_of_movies_user_watched = set()
+    for movie in user_data["watched"]:
+        titles_of_movies_user_watched.add(movie["title"])
```
Comment (anselrognlie) [inline viewing_party/party.py:123]: This would be another great spot to try out a comprehension!
```python
    titles_of_movies_user_watched = set(movie["title"] for movie in user_data["watched"])
```

File: viewing_party/party.py [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+    for movie in friends_movies:
+        if movie["title"] not in titles_of_movies_user_watched:
+            if in_users_subcriptions(movie["host"], user_data["subscriptions"]):
```
Comment (anselrognlie) [inline viewing_party/party.py:127]: Here's what I mentioned where the helper function name might actually make this less readbale, since this is all we need here.
```python
            if movie["host"] in user_data["subscriptions"]:
```

File: viewing_party/party.py [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+                movie_recs.append(movie)
+    return movie_recs
+
+# Wave 5
+def get_new_rec_by_genre(user_data):
+
+    genre_list = [a_dict['genre'] for a_dict in user_data['watched']]
+    most_common_genre = max(genre_list, key = genre_list.count, default = 0)
```
Comment (anselrognlie) [inline viewing_party/party.py:135]: Don't you already have a function that calculates the user's most common genre?

File: viewing_party/party.py [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+    friends_movies = get_friends_movies(user_data)
+
+    titles_of_movies_user_watched = set()
+    for movie in user_data["watched"]:
+        titles_of_movies_user_watched.add(movie["title"])
+
+    movie_recs = []
+    for movie in friends_movies:
+        if movie["title"] not in titles_of_movies_user_watched:
+            if most_common_genre in movie.values():
```
Comment (anselrognlie) [inline viewing_party/party.py:146]: If we know exactly which key we need to check, no need to iterate over the values.
```python
            if most_common_genre == movie["genre"]:
```

File: viewing_party/party.py [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+    titles_of_movies_user_watched = set()
+    for movie in user_data["favorites"]:
+        titles_of_movies_user_watched.add(movie["title"])
+
+    movie_recs = []
+    for recommended_movies in user_data['favorites']:
+        if recommended_movies not in friends_movies:
+            movie_recs.append(recommended_movies)
+    
+    if movie_recs == False:
```
Comment (anselrognlie) [inline viewing_party/party.py:166]: `movie_recs` is initialized to at least an empty list, so it will never be False. Remember that we can evaluate a list in a boolean context to see whether it is truthy (non-empty) or falsy (empty), but that's not the same as equaling True or False.

File: viewing_party/party.py [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+    [movie_recs_no_dups.append(x) for x in movie_recs if x not in movie_recs_no_dups]
+    return movie_recs_no_dups
+
+def get_rec_from_favorites(user_data):
+
+    friends_movies = get_friends_movies(user_data)
+
+    titles_of_movies_user_watched = set()
+    for movie in user_data["favorites"]:
+        titles_of_movies_user_watched.add(movie["title"])
```
Comment (anselrognlie) [inline viewing_party/party.py:159]: It doesn't look like this set is being used. But maybe if you gathered the titles of the movies the _friends_ have watched instead, it could help the following loop!

**Review by anselrognlie — COMMENTED** [review]
Nice job!

All waves are complete and all tests are passing! Your code is well organized, and you are generally using good, descriptive variable names. You are making good choices about which loops to use, and which helper data structures to assist you. Happy to see that you were thinking about adding additional helper methods too!

It looks like you're starting to stretch your legs a bit with python list comprehensions and using additional library features, which is great as well! Just make sure that you are using them as intended (e.g. list comprehensions make lists, so if you don't need the resulting list, probably stick with a regular loop or modify the comprehension). And be sure to have a good understanding of what the library routines are doing. And early in our learning, skipping the library function and doing our own implementation can still be really valuable!

But overall, great job!
